File: simucore_pytest/core/application_tree.py
# noqa: N815
from pydantic import BaseModel


class ConnectedInput(BaseModel):
    id: int

class Parameter(BaseModel):
    id: int
    name: str
    typeName: str
    value: str

class Output(BaseModel):
    id: int
    name: str
    typeName: str
    value: str
    connectedInputs: list[ConnectedInput] | None = []

class Input(BaseModel):
    id: int
    name: str
    typeName: str
    value: str

class PhysicalInput(Input):
    pass

class PhysicalOutput(Output):
    pass

class Component(BaseModel):
    id: int
    name: str
    Components: list[Component] | None = []
    Inputs: list[Input] | None = []
    Outputs: list[Output] | None = []
    PhysicalOutputs: list[PhysicalOutput] | None = None
    PhysicalInputs: list[PhysicalInput] | None = None
    Parameters: list[Parameter] | None = []

Component.model_rebuild()

class ApplicationTree(BaseModel):
    id: int
    name: str
    Components: list[Component]
    config: dict = {}

File: src/generic/ApplicationTree.cpp
#include <SimuCore/ApplicationTree.hpp>
#include <SimuCore/Signal.hpp>
#include <SimuCore/generated/Config.hpp>


const nlohmann::json ApplicationTree::getApplicationTreeAsJson()
{
    auto buildJsonTree = [](Component *component, auto &self) -> nlohmann::json
    {
        nlohmann::json componentJson;
        componentJson["name"] = component->getName();
        componentJson["id"] = component->getId();
        
        if (component->getComponentType() == ComponentType::INTERNAL_INPUT ||
            component->getComponentType() == ComponentType::INTERNAL_OUTPUT ||
            component->getComponentType() == ComponentType::PHYSICAL_INPUT ||
            component->getComponentType() == ComponentType::PHYSICAL_OUTPUT ||
            component->getComponentType() == ComponentType::PARAMETER)
        {
            const auto signal = SignalRegistry::getInstance().find(component->getId());

            componentJson["value"] = signal->getValueAsString();
            componentJson["typeName"] = signal->getTypeName();
            
            if (component->getComponentType() == ComponentType::INTERNAL_OUTPUT)
            {
                for (auto *connectedInput : signal->getConnectedBaseSignals())
                {
                    if (componentJson.find("connectedInputs") == componentJson.end())
                    {
                        componentJson["connectedInputs"] = nlohmann::json::array();
                    }
                    componentJson["connectedInputs"].push_back(
                        nlohmann::json{{"id", connectedInput->getId()}});
                }
            }
        }

        auto subComponents = component->getSubComponents();
        if (!subComponents.empty())
        {
            for (auto *subComponent : subComponents)
            {
                if (subComponent)
                {
                    std::string componentType = subComponent->getComponentTypeName() + "s";
                    if (componentJson.find(componentType) == componentJson.end())
                    {
                        componentJson[componentType] = nlohmann::json::array();
                    }
                    componentJson[componentType].push_back(self(subComponent, self));
                }
            }
        }

        return componentJson;
    };
    nlohmann::json applicationTree = buildJsonTree(_root, buildJsonTree);
    return applicationTree;
}